### PR TITLE
[HYDRATOR-1074] Implementation for HTTP To HDFS Action plugin

### PR DESCRIPTION
**Pull request by shashankmoghe**
Design : https://wiki.cask.co/display/CE/HTTPToHDFS+Action
JIRA: https://issues.cask.co/browse/HYDRATOR-1074

### CODE REVIEW

File: LICENSE.txt [hunk 1]
```diff
@@ -0,0 +1,232 @@
+Copyright © 2014-2015 Cask Data, Inc.
```
Comment (russorat) [inline LICENSE.txt]: 2017

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,78 @@
+=================
+HTTPTOHDFS Plugin
```
Comment (russorat) [inline README.rst]: HTTPToHDFS

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+Introduction
+============
+Action plugin to fetch data from an external http webservice or url and dump the data into HDFS.
+
+Getting Started
+===============
+
+Prerequisites
+--------------
+CDAP version 3.5.x or higher.
```
Comment (russorat) [inline README.rst]: 4.0 or higher? unless you tested on a 3.5 cluster.

File: docs/HTTPToHDFS-action.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
+# HTTP To HDFS Action
+
+
+Description
+-----------
+Action to fetch data from an external http webservice or url and dump the data into HDFS.
```
Comment (russorat) [inline docs/HTTPToHDFS-action.md]: ...from an external http endpoint and create a file in HDFS.

File: docs/HTTPToHDFS-action.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+Use Cases
+--------
```
Comment (russorat) [inline docs/HTTPToHDFS-action.md]: length should match line above

File: docs/HTTPToHDFS-action.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
+# HTTP To HDFS Action
+
+
+Description
+-----------
+Action to fetch data from an external http webservice or url and dump the data into HDFS.
+
+Use Cases
```
Comment (russorat) [inline docs/HTTPToHDFS-action.md]: remove the use cases section please

File: docs/HTTPToHDFS-action.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+and write the results to the /tmp/ dir in HDFS so that the rest of the pipeline can process the data.
+3.I have partnered with a 3rd party company that is sending me a large amount of data in a gzip file.
+The file is stored on a webserver somewhere and i am given a url to download it.
+Using this plugin, i configure it to pull the binary file, store the .gz file on hdfs, and use the file source to natively read that data for processing.
+
+
+Properties
+----------
+
+**hdfsFilePath:** The location to write the data in HDFS. If the file already exists,same will be overwritten.
```
Comment (russorat) [inline docs/HTTPToHDFS-action.md]: file already exists, it will be overwritten.

File: docs/HTTPToHDFS-action.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+**url:** The URL to fetch data from.
+
+**method:** The HTTP request method.GET and POST are the allowed methods.
```
Comment (russorat) [inline docs/HTTPToHDFS-action.md]: space after .

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 9]
```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java]: 2017

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509TrustManager;
+import javax.ws.rs.HttpMethod;
+
+/**
+ * Action to fetch data from an external http webservice or url and dump the data into HDFS.
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java]: ...from an external http endpoint and create a file in HDFS.

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+ */
+
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name("HTTPToHDFS")
+@Description("Action to fetch data from an external http webservice or url and dump the data into HDFS.")
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java]: from an external http endpoint and create a file in HDFS.

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+    int retries = 0;
+    Exception exception = null;
+    do {
+      HttpURLConnection conn = null;
+      Map<String, String> headers = config.getRequestHeadersMap();
+      try {
+        URL url = new URL(config.url);
+        conn = (HttpURLConnection) url.openConnection();
+        if (conn instanceof HttpsURLConnection) {
+          System.setProperty("https.protocols", "TLSv1,TLSv1.1,TLSv1.2");
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java:95]: could you add a comment as to why this is required?

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+            }
+          }
+          context.getArguments().set(config.responseHeaders, responseHeader.toString());
+        }
+        break;
+      } catch (MalformedURLException | ProtocolException e) {
+        throw new IllegalStateException("Error opening url connection. Reason: " + e.getMessage(), e);
+      } catch (Exception e) {
+        LOG.warn("Error making {} request to url {} with headers {}.", config.method, config.url, headers);
+        exception = e;
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java:151]: should this fail the pipeline?

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+            }
+          }
+          context.getArguments().set(config.outputPath, config.hdfsFilePath);
+          StringBuilder responseHeader = new StringBuilder();
+          for (Map.Entry<String, List<String>> k : conn.getHeaderFields().entrySet()) {
+            if (!Strings.isNullOrEmpty(k.getKey())) {
+              responseHeader.append(k.toString()).append("\n");
+            }
+          }
+          context.getArguments().set(config.responseHeaders, responseHeader.toString());
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java]: Shouldn't this append the key and value to the response headers? could you provide an example output of what the string that would be stored for the response headers?

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+    };
+    HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
+  }
+
+  /**
+   * Config for the http to hdfs action.
+   */
+  public static class HTTPToHDFSActionConfig extends PluginConfig {
+
+    @Description("The location to write the data in HDFS. If the file already exists,same will be overwritten.")
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java]: space after ,

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+      this.numRetries = numRetries;
+      this.readTimeout = readTimeout;
+      this.connectTimeout = connectTimeout;
+      this.followRedirects = followRedirects;
+      this.disableSSLValidation = disableSSLValidation;
+      this.outputPath = outputPath;
+      this.responseHeaders = responseHeaders;
+    }
+
+    public HTTPToHDFSActionConfig() {
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java]: could you move this above the other constructor?

File: src/test/java/co/cask/hydrator/plugin/batch/HTTPToHDFSActionTest.java [hunk 17]
```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
Comment (russorat) [inline src/test/java/co/cask/hydrator/plugin/batch/HTTPToHDFSActionTest.java]: 2017

File: src/test/java/co/cask/hydrator/plugin/mock/MockFeedHandler.java [hunk 18]
```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
```
Comment (russorat) [inline src/test/java/co/cask/hydrator/plugin/mock/MockFeedHandler.java]: 2017

File: widgets/HTTPToHDFS-action.json [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+      "label": "HTTP To HDFS Action Properties",
+      "properties": [
+        {
+          "widget-type": "textbox",
+          "label": "URL",
+          "name": "url"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "The Location To Write The Data In HDFS",
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: Shorten to "HDFS File Path"

File: widgets/HTTPToHDFS-action.json [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+          "name": "connectTimeoutMillis"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Read Timeout (milliseconds)",
+          "name": "readTimeoutMillis"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Output File Path For The Data That Was Written",
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: Change to "Token Key for HDFS File Path"

File: widgets/HTTPToHDFS-action.json [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+          "name": "outputPath"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Response Headers",
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: Token Key for Response Headers

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java (hunk 14 above)
Comment (russorat) [reply src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java]: Also, it is probably a good idea to have the response headers in a JSON formatted string, so they can be used later in the pipeline. Something like { "header1" : "value1", ... } as a String field.

File: widgets/HTTPToHDFS-action.json [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+              "UTF-16",
+              "UTF-16BE",
+              "UTF-16LE"
+            ],
+            "default": "UTF-8"
+          }
+        },
+        {
+          "widget-type": "select",
+          "label": "Should Follow Redirects",
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: End with ?

File: widgets/HTTPToHDFS-action.json [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+        {
+          "widget-type": "textbox",
+          "label": "Output File Path For The Data That Was Written",
+          "name": "outputPath"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Response Headers",
+          "name": "responseHeaders"
+        }
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json:150]: Missing entries for readTimeout and connectTimeout

File: widgets/HTTPToHDFS-action.json [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Read Timeout (milliseconds)",
+          "name": "readTimeoutMillis"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Output File Path For The Data That Was Written",
+          "name": "outputPath"
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: Also, add the default value here

File: widgets/HTTPToHDFS-action.json [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Response Headers",
+          "name": "responseHeaders"
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: and the default value here

File: widgets/HTTPToHDFS-action.json (hunk 23 above)
Comment (shashankmoghe) [reply widgets/HTTPToHDFS-action.json:150]: These entries are present.Please refer line 120 and 125 in json

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java (hunk 13 above)
Comment (shashankmoghe) [reply src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java:151]: To manage retry logic,Exception is not thrown from here, instead thrown after the while loop.Implementation is similar to HTTPCallbackAction.

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java (hunk 12 above)
Comment (shashankmoghe) [reply src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java:95]: Sure.This is required to disable SSLv3.Will add comment.

File: widgets/HTTPToHDFS-action.json (hunk 23 above)
Comment (shashankmoghe) [reply widgets/HTTPToHDFS-action.json:150]: Oh,got it,Names were different.Will update in PR.

**Comment by shashankmoghe** [comment]
@russorat Incorporated all the review comments.Please check.

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java (hunk 13 above)
Comment (russorat) [reply src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java:151]: ok. but what happens if it fails the first time, but succeeds on the second? I don't see where you set the exception to null again after a successful run. which means line 160 will still happen, even if a retry succeeded.

File: widgets/HTTPToHDFS-action.json [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+          "widget-type": "textbox",
+          "label": "Token Key for HDFS File Path",
+          "name": "outputPath",
+          "default": "filePath"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Token Key for Response Headers",
+          "name": "responseHeaders",
+          "default": "responseHeaders"
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: this isn't correct. the proper way for this is "widget-attributes" : { "default" : "responseHeaders" }

It's clear you didn't test this change.

File: widgets/HTTPToHDFS-action.json [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Connection Timeout (milliseconds)",
+          "name": "connectTimeout"
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Read Timeout (milliseconds)",
+          "name": "readTimeout"
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: add default

File: widgets/HTTPToHDFS-action.json [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+              "9",
+              "10"
+            ],
+            "default": "3"
+          }
+        },
+        {
+          "widget-type": "textbox",
+          "label": "Connection Timeout (milliseconds)",
+          "name": "connectTimeout"
```
Comment (russorat) [inline widgets/HTTPToHDFS-action.json]: add default

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 29]
```diff
@@ -0,0 +1,344 @@
...
+
+    @Description("The number of times the request should be retried if the request fails. Defaults to 3.")
+    private Integer numRetries;
+
+    @Description("Sets the connection timeout in milliseconds. Set to 0 for infinite. Default is 60000 (1 minute).")
+    @Nullable
+    private Integer connectTimeout;
+
+    @Description("The time in milliseconds to wait for a read. Set to 0 for infinite. Defaults to 60000 (1 minute).")
+    @Nullable
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java:247]: this should be macro enabled

File: src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java [hunk 30]
```diff
@@ -0,0 +1,344 @@
...
+    @Description("If user enables SSL validation, they will be expected to add the certificate to the trustStore" +
+      " on each machine. Defaults to true.")
+    private Boolean disableSSLValidation;
+
+    @Description("The number of times the request should be retried if the request fails. Defaults to 3.")
+    private Integer numRetries;
+
+    @Description("Sets the connection timeout in milliseconds. Set to 0 for infinite. Default is 60000 (1 minute).")
+    @Nullable
+    private Integer connectTimeout;
```
Comment (russorat) [inline src/main/java/co/cask/hydrator/plugin/batch/HTTPToHDFSAction.java:244]: this should be macro enabled

File: docs/HTTPToHDFS-action.md [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+
+Example
+-------
+This example performs HTTP GET request to http://example.com/data and downloads the csv file to file:///tmp/data.csv.
+
+    {
+        "name": "HTTPToHDFSAction",
+        "type": "action",
+        "properties": {
+            "hdfsFilePath": "file://tmp/data.csv",
```
Comment (russorat) [inline docs/HTTPToHDFS-action.md]: you don't need the file:// on this. /tmp/data.csv is fine

File: README.rst [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+
+IRC Channel
+-----------
+CDAP IRC Channel: #cdap on irc.freenode.net
+
+
+License and Trademarks
+======================
+
+Copyright © 2015-2016 Cask Data, Inc.
```
Comment (russorat) [inline README.rst]: 2017

**Comment by russorat** [comment]
LGTM

**Pull request merged**